app/uk/gov/hmrc/gform/obligation/ObligationService.scala

```diff
@@ -17,30 +17,114 @@
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
     val hmrcTaxPeriodIdentifiers = formTemplate.expandFormTemplateFull.allFCs.collect {
       case IsHmrcTaxPeriod(el) => el
     }
-    val futureListOfTaxPeriodDes =
-      gformConnector
-        .getAllTaxPeriods(hmrcTaxPeriodIdentifiers)
-        .map(i => i.flatMap(j => j.obligation.obligations.flatMap(h => makeMap(j.id, h.obligationDetails))))
-    futureListOfTaxPeriodDes.map(i => i.toMap)
+    for {
+      string <- authService.evaluateSubmissionReference(
+                 hmrcTaxPeriodIdentifiers.head.idNumber,
+                 retrievals,
+                 formTemplate,
+                 FormDataHelpers.formDataMap(form.formData),
+                 form.envelopeId)
+      output <- form.obligations match {
+                 case Some(x) => {
+                   for {
+                     idNumbers <- Future.sequence(
+                                   x.listAllInfo.map(
+                                     i =>
+                                       authService.evaluateSubmissionReference(
+                                         i.idNumber,
+                                         retrievals,
+                                         formTemplate,
+                                         FormDataHelpers.formDataMap(form.formData),
+                                         form.envelopeId)))
+                     output <- {
+                       if (idNumbers.forall(i => string.equals(i))) {
+                         Future(form)
+                       } else {
+                         val newObligations = lookupObligationsMultiple(formTemplate, authService, retrievals, form)
+                         newObligations.map(i => form.copy(obligations = Some(ListAllInfo(i))))
+                       }
+                     }
+                   } yield output
+                 }
+                 case None => {
+                   if (hmrcTaxPeriodIdentifiers.forall(i => !i.regimeType.value.isEmpty && !string.isEmpty)) {
+                     val newObligations = lookupObligationsMultiple(formTemplate, authService, retrievals, form)
+                     newObligations.map(i => form.copy(obligations = Some(ListAllInfo(i))))
+                   } else {
+                     Future(form)
+                   }
+                 }
+               }
+    } yield output
   }
 
-  def makeMap(id: HmrcTaxPeriod, obligation: List[ObligationDetail]) =
-    Map(
-      id ->
-        TaxPeriods(obligation.map(l =>
-          TaxPeriod(l.inboundCorrespondenceFromDate, l.inboundCorrespondenceToDate, l.periodKey))))
+  def updateObligations(formId: FormId, userData: UserData, form: Form, newForm: Form)(
+    implicit hc: HeaderCarrier,
+    ec: ExecutionContext) =
+    if (form.obligations != newForm.obligations) {
+      gformConnector.updateUserData(formId, userData)
+    } else {
+      Future.successful()
```

Comment:
What are we returning here ???

Reply:
The idea was that as updateUserData returns a Future[Unit] the else needed to return the same

Reply:
The thing is you are not passing `()` as the parameter to the `successful` method, do you see it?